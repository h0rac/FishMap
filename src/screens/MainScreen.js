import React, { Component, PropTypes } from 'react';
import FishMarker from '../components/FishMarker';
import UserMarker from '../components/UserMarker';

import {
	StyleSheet,
	View,
	Alert,
	AsyncStorage,
	Text,
	ActivityIndicator
} from 'react-native';
import MapView from 'react-native-maps';

import { setFishmark, setMapViewForAnimation, shareMyWaypoint } from '../actions/fishmarks';
import { deleteFishmarkPosition } from '../actions/fishmarks';
import { connect } from 'react-redux';
import {Icon} from 'react-native-elements';
import { loadFishPositions, IOsetFishmarksCandidateList } from '../actions/fishmarks';
import {
	logout, getUserLocation, setIntervalID, checkAuthToken, setIntervalAlive
} from '../actions/user';

import { displayAlert } from '../common/utils';


const labelStyle = (props, alignSelf, marginTop) => ({
	fontSize: 14,
	fontWeight: '500',
	marginTop,
	color: props.focused ? props.tintColor : 'white'
});

let timeoutID;
let intervalId;

let _mapView = MapView;

class MainScreen extends Component {

	constructor() {
		super();

		this.handleFishMarkerSet = this.handleFishMarkerSet.bind(this);
		this._handleFishMarkPress = this._handleFishMarkPress.bind(this);
		this.handleLogout = this.handleLogout.bind(this);
		this.logoutUser = this.logoutUser.bind(this);
		this.onReceiveFishmark = this.onReceiveFishmark.bind(this);
		this.setInitialUserPosition = this.setInitialUserPosition.bind(this);
		this.onReceiveError = this.onReceiveError.bind(this);
		this.handleSharing = this.handleSharing.bind(this);

		this.state = {
			interval: 0,
			intervalId: 0,
			test: null,
			clear: null,
			intervalAlive: true
		};
	}

	onReceiveError = (data) => {
		displayAlert('Authentication', data);
		this.props.socketIO.emit('onErrorDisconnect');
		this.props.navigation.navigate('LoginScreen');

	};

	logoutUser() {
		AsyncStorage.getItem('token').then(token => {
			if (token) {
				this.props.socketIO.emit('onForceDisconnect', { token: JSON.parse(decodeURI(token)) });
			}
		});
		clearInterval(this.props.timeoutID);
		this.props.dispatch(setIntervalAlive(true));
		this.props.dispatch(logout());
		this.props.navigation.navigate('LoginScreen');

	}

	handleLogout() {
		Alert.alert(
			'Logout',
			'Do you want to logout from FishMap ?',
			[
				{ text: 'OK', onPress: () => this.logoutUser(), style: 'ok' },
				{ text: 'Cancel', onPress: () => console.log('cancel'), style: 'cancel' }
			],
			{ cancelable: true }
		);
	};

	static navigationOptions = ({ navigation }) => {
		const { params = {} } = navigation.state;


		return {

			headerStyle: {
				backgroundColor: '#2F95D6'
			},
			headerTintColor: 'white',
			tabBarLabel: (props) => (<Text style={labelStyle(props, 'flex-end', 15)}> Fish Map </Text>),
			headerLeft: <Icon
				name="menu"
				size={28}
				color={'white'}
				containerStyle={{ paddingLeft: 10 }}
				underlayColor={"#2F95D6"}
				onPress={() => navigation.navigate('DrawerOpen')}
			/>,
			headerRight:
				<Icon name="exit-to-app"
				      size={28}
				      color={'white'}
				      containerStyle={{ paddingRight: 10 }}
				      underlayColor={"#2F95D6"}
				      onPress={() => params.handleLogout()}/>,
			tabBarIcon:
				<Icon
					name="home"
					size={26}
					color={'white'}
				/>
		};
	};


	componentWillMount() {
		this.setState({
			interval: this.props.duration,
			timeoutID: this.props.timeoutID,
			intervalAlive: this.props.intervalAlive
		});
	}


	componentDidMount() {
		AsyncStorage.getItem('token').then(token => {
			if (!token) {
				this.props.navigation.navigate('LoginScreen');
			}
			else {
				this.props.dispatch(checkAuthToken(this.props.navigation, 'MainScreen'));
				if (this.props.intervalAlive) {
					clearInterval(this.props.timeoutID);
					intervalId = setInterval(() => this.props.socketIO.emit('onFishmarkUpdate', {
						token: JSON.parse(decodeURI(token)),
						receive: this.props.receive
					}), this.state.interval);
					this.props.dispatch(setIntervalID(intervalId));
				}
				this.props.dispatch(getUserLocation());
				this.props.dispatch(loadFishPositions());
				this.props.navigation.setParams({ handleLogout: this.handleLogout });
				this.props.socketIO.on('onReceiveFishmark', data => this.onReceiveFishmark(data));
				this.props.socketIO.on('onReceiveError', data => this.onReceiveError(data.error));
			}
		});
	}


	onReceiveFishmark = (data) => {
		data.waypoints && !data.pause ? this.props.dispatch(IOsetFishmarksCandidateList(data.waypoints)) : null;
	};


	handleFishMarkerSet(e) {

		this.props.dispatch(setFishmark({
			...e.nativeEvent.coordinate, latitudeDelta: 0.0922,
			longitudeDelta: 0.0421, title: 'Marker-' + e.nativeEvent.coordinate.latitude.toFixed(6),
			key: new Date().getTime(), date: new Date().toDateString()
		}));
	};

	handleSharing = (position) => {
		const toShare = this.props.positions.find(waypoint => waypoint.latitude === position.latitude && waypoint.longitude === waypoint.longitude);
		this.props.dispatch(shareMyWaypoint(toShare));
		this.props.navigation.navigate('SharingScreen');
	};


	_handleFishMarkPress = (e) => {
		const position = e.nativeEvent.coordinate;

		Alert.alert(
			'Waypoint',
			'What do you want to do with Waypoint ?',
			[
				{ text: 'Share', onPress: () => this.handleSharing(position) },
				{ text: 'Edit', onPress: () => this.props.navigation.navigate('WayPointEditScreen') },
				{ text: 'Delete', onPress: () => this.props.dispatch(deleteFishmarkPosition(position)) },
				{ text: 'Cancel', onPress: () => console.log('cancel'), style: 'cancel' }
			],
			{ cancelable: true }
		);
	};

	setInitialUserPosition = () => {
		let initPosition;
		const userPos = this.props.userLocation.latitude ? this.props.userLocation : null;
		if (userPos || this.props.positions.length === 0) {
			initPosition = userPos;
		}
		return initPosition;
	};


	render() {

		const initUserPosition = this.setInitialUserPosition();
		const loading = this.props.isFetching;
		return (
			<View style={styles.container}>

				<MapView onLongPress={this.handleFishMarkerSet}
					//onRegionChangeComplete={()=> this.props.dispatch(loadFishmarkPositions())}
					       ref={(mapView) => {
						       _mapView = mapView;
						       this.props.dispatch(setMapViewForAnimation(_mapView));
					       }}
					       style={styles.map}
					       initialRegion={{
						       latitude: 54.475408,
						       longitude: 18.263086,
						       latitudeDelta: 0.0922,
						       longitudeDelta: 0.0421
					       }}>

					{this.props.positions.map((marker, index) =>
						<FishMarker key={index} marker={marker}
						            callbackPress={this._handleFishMarkPress}
						/>
					)}
					{this.props.userLocation.latitude ?
						<UserMarker marker={initUserPosition}/> : null}
				</MapView>
				{loading === true ? <ActivityIndicator size={'large'}/> : null}
			</View>
		);
	}
}


const styles = {
	container: {
		flex: 1,
		justifyContent: 'center',
		alignItems: 'center'

	},
	notificator: {
		flex: 1,
		flexDirection: 'column',
		alignItems: 'flex-end',
		marginRight: 200
	},
	button: {
		alignItems: 'center',
		backgroundColor: '#2F95D6',
		height: 40,
		padding: 20
	},

	map: {
		...StyleSheet.absoluteFillObject
	}
};

const mapStateToProps = state => {
	return {
		positions: state.fishmarks.fishmarks,
		selectedPosition: state.fishmarks.region,
		isSelected: state.fishmarks.selected,
		isFetching: state.fishmarks.isFetching,
		userLocation: state.user.position,
		sharedFishmarks: state.fishmarks.sharedFishmarks,
		receive: state.user.receive,
		duration: state.user.duration,
		emitStatus: state.user.emitStatus,
		socketIO: state.user.socketIO,
		timeoutID: state.user.timeoutID,
		intervalAlive: state.user.intervalAlive
	};
};

export default connect(mapStateToProps)(MainScreen);



