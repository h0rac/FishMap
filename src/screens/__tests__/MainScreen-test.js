import { TextInput, Keyboard, Dimensions, Platform, Alert, AsyncStorage, StyleSheet } from 'react-native';
import React from 'react';
import MainScreen from '../MainScreen';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';


jest.mock('Alert', () => {
	return {
		alert: jest.fn()
	};
});


jest.mock('AsyncStorage', () => {
	return {
		getItem: jest.fn((item) => {
			return new Promise((resolve, reject) => {
				resolve('"1234"');
			});
		})
	};
});


jest.useFakeTimers();

configure({ adapter: new Adapter() });

const middlewares = [];
const mockStore = configureStore(middlewares);

jest.mock('react-native-vector-icons/Ionicons', () => ({}));

jest.mock('react-native-vector-icons/FontAwesome', () => ({}));


describe('Testing Main screen', () => {

	let tree;
	let navigate;
	let state;
	let props;
	let dispatch;

	const event = {
		nativeEvent: {
			coordinate: {
				latitude: 54.475408,
				longitude: 18.263086,
				latitudeDelta: 0.0922,
				longitudeDelta: 0.0421
			}
		}
	};

	beforeEach(() => {
		navigate = jest.fn();
		state = {
			fishmarks: { fishmarks: [] },
			user: { position: {}, duration: 1000 }

		};
		const store = mockStore(state);
		dispatch = jest.fn();

		props = store;

		tree = shallow(
			<MainScreen {...props} store={store} navigation={{ navigate }}
			/>, { lifecycleExperimental: true });
	});

	it('renders correctly', () => {
		expect(tree.dive()).toMatchSnapshot();
	});


	it('should set user initial position', () => {
		const MainScreenComponent = tree.dive();
		MainScreenComponent.setProps({
			userLocation: {
				latitude: 54.475408,
				longitude: 18.263086,
				latitudeDelta: 0.0922,
				longitudeDelta: 0.0421
			}
		});
	});

	it('should invoke alert window on fishmark press', () => {
		const MainScreenComponent = tree.dive();
		const handleFishMarkPress = jest.spyOn(MainScreenComponent.instance(), '_handleFishMarkPress');
		handleFishMarkPress(event);
		expect(handleFishMarkPress).toBeCalledWith(event);
	});

	it('should invoke cancel on fishmark alert window', () => {
		Alert.alert.mock.calls[0][2][2].onPress();
	});


	it('should invoke delete on fishmark when user select delete from alert window', () => {
		Alert.alert.mock.calls[0][2][1].onPress();
	});

	it('should move to edit screen when user select edit from alert window', () => {
		Alert.alert.mock.calls[0][2][0].onPress();
		dispatch(navigate('WayPointEditScreen'));
		expect(dispatch).toHaveBeenCalled();
	});


	it('should set fishmark position when long press on Map', () => {
		const MainScreenComponent = tree.dive();
		const MapView = MainScreenComponent.find('MapView');
		MapView.first().simulate('LongPress', event);
	});


	it('should invoke onReceive using IO Socket with pause true', () => {
		const MainScreenComponent = tree.dive();
		const onReceive = jest.spyOn(MainScreenComponent.instance(), 'onReceiveFishmark');
		const data = { pause: true };
		onReceive(data);
		expect(onReceive).toHaveBeenCalledWith(data);
	});

	it('should invoke onReceive using IO Socket with pause false and waypoint available', () => {
		const MainScreenComponent = tree.dive();
		const onReceive = jest.spyOn(MainScreenComponent.instance(), 'onReceiveFishmark');
		const data = {
			waypoints: [{
				latitude: 54.475408,
				longitude: 18.263086,
				latitudeDelta: 0.0922,
				longitudeDelta: 0.0421
			}]
		};
		onReceive(data);
	});


	it('should dispatch loadFishPositions action', () => {
		const MainScreenComponent = tree.dive();
		const componentDidMount = jest.spyOn(MainScreenComponent.instance(), 'componentDidMount');
		componentDidMount();
	});


	it('should have user token in AsyncStorage', () => {
		const MainScreenComponent = tree.dive();
		const componentDidMount = jest.spyOn(MainScreenComponent.instance(), 'componentDidMount');
		componentDidMount();
		expect(componentDidMount).toBeCalled();

		spyOn(MainScreenComponent.instance(), 'componentDidMount').and.returnValue(function () {
			return AsyncStorage.getItem('token');
		});
		expect(MainScreenComponent.instance().state.interval).not.toBe(null);
	});

	it('should not have token in AsyncStorage and redirect to LoginScreen', () => {
		const MainScreenComponent = tree.dive();

		spyOn(MainScreenComponent.instance(), 'componentDidMount').and.returnValue(function () {
			return AsyncStorage.getItem('fake');
		});
	});


});