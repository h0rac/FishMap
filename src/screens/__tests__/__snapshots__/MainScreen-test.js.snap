// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main screen renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainScreen
    clearActions={[Function]}
    dispatch={[Function]}
    duration={1000}
    getActions={[Function]}
    getState={[Function]}
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
    positions={Array []}
    replaceReducer={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    subscribe={[Function]}
    userLocation={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <MapView
          initialRegion={
            Object {
              "latitude": 54.475408,
              "latitudeDelta": 0.0922,
              "longitude": 18.263086,
              "longitudeDelta": 0.0421,
            }
          }
          onLongPress={[Function]}
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />,
        null,
      ],
      "style": Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            Array [],
            null,
          ],
          "initialRegion": Object {
            "latitude": 54.475408,
            "latitudeDelta": 0.0922,
            "longitude": 18.263086,
            "longitudeDelta": 0.0421,
          },
          "onLongPress": [Function],
          "style": Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        },
        "ref": [Function],
        "rendered": Array [
          null,
        ],
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MapView
            initialRegion={
              Object {
                "latitude": 54.475408,
                "latitudeDelta": 0.0922,
                "longitude": 18.263086,
                "longitudeDelta": 0.0421,
              }
            }
            onLongPress={[Function]}
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />,
          null,
        ],
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              Array [],
              null,
            ],
            "initialRegion": Object {
              "latitude": 54.475408,
              "latitudeDelta": 0.0922,
              "longitude": 18.263086,
              "longitudeDelta": 0.0421,
            },
            "onLongPress": [Function],
            "style": Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          },
          "ref": [Function],
          "rendered": Array [
            null,
          ],
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
    "lifecycleExperimental": true,
  },
}
`;
